### PR TITLE
Continue implementation of Solidity bindings rules in the new binder

### PR DESCRIPTION
**Pull request by ggiraldez**
This PR implements most bindings rules in order to bring feature parity with the existing stack graph binder, at least as far as the internal bindings snapshots concern.

This mainly entails completing the implementation of the definition typing and reference resolution passes. The final current pipeline consists of 4 passes:

- `p2_collect_definitions` in which all definitions are collected with their naming identifiers. Also lexical (and other kinds of) scopes are identified and linked together, and definitions are registered into them for later lookup.
- `p3_linearise_contracts` which will collect all bases of contracts and interfaces and then compute the linearisation for each of them.
- `p4_type_definitions` which will determine and assign typing information to all the definitions collected in `p2` and as a result register most types used in the compilation unit. Additionally, public state variables will be assigned an equivalent getter function type, which happens after the main typing pass to ensure all types are already registered.
- `p5_resolve_references` will find identifiers used as references, resolve them to the appropriate definitions, and compute typing information for AST nodes containing expressions and statements. Both these actions are co-dependant and happen concurrently for each node.

All the information collected and computed by these passes are stored in a `Binder` and `TypeRegistry` instances.

The `Binder` will store:
- `Definition` objects (identifier + definiens node), indexed by definiens `NodeId`
- `Reference` objects (identifier + `Resolution`), indexed by identifier `NodeId`
- `Scope` objects (set of definitions and relationship to other scopes), indexed by `NodeId`
- index of root `FileScope`s for all files in the compilation unit
- `UsingDirective` objects registered both at valid scopes (contracts/libraries/etc) and globally
- linearisations, as a vector of definitions, indexed by the contract/interface definition's `NodeId`
- `Typing` information for each relevant `NodeId`

The `TypeRegistry` stores an index of registered types, both elementary types and user defined types. Each type is given a `TypeId` for efficient storage, lookup and comparison. The registry also stores some common elementary types required to type most built-ins functions and some kinds of expressions (eg. the boolean type).

There are two important value types that are central to how binding works, and are closely related: `Resolution` and `Typing`.

`Resolution` represents the result of a lookup of a reference in the context where that reference occurs. This can be:
- `Unresolved`, the identifier was not found.
- `Definition` to a single definition identified by a `NodeId`.
- `Ambiguous` when the lookup returns multiple possible definitions (eg. multiple function overloads, or virtual methods in a contract hierarchy). It's usually not possible to determine *which* of the found definitions should apply, as more information is needed. Eg. the types of arguments when the reference is used in a function call, to select the appropriate overload.
- `BuiltIn` which means the symbol refers to a Solidity built-in of some kind. The possible variants are encoded in an `enum` and the behaviour of each is encoded in the `built_ins.rs` module.

`Typing` represents typing information related to an AST node. If the node happens to be a reference identifier, there is a direct relationship between its `Resolution` and the `Typing`. The possible variants are:
- `Unresolved` the type of the node cannot be resolved at this time (this normally propagates upwards).
- `Resolved` with a corresponding `TypeId` valid in the `TypeRegistry`.
- `Undetermined` when there are multiple possible types that result from the typing of an `Ambiguous` resolution. Function call overload resolution will select the appropriate type by matching the types of the arguments and work backwards to the reference (if any) and fixup the selected definition.
- `UserMetaType` is a node which refers to a user defined type by name, eg. the identifier of a contract. This is not a type that can be translated to the EVM, and as such is a way to suspend the typing until more information on how it's used is gathered. Eg. to create a new contract via the `new` operator, or construct a struct by using the struct's name in a function call.
- `MetaType` similar to `UserMetaType`, but refers to the meta type of an elementary type. A typical use case is explicit casting, which parses as a function call.
- `BuiltIn` is used to type the `BuiltIn` resolution when the result is not yet an actual type representable in the EVM. Eg. the built-in function `addmod` resolves to `Resolution::BuiltIn(Addmod)` and will type to `Typing::BuiltIn(Addmod)` which in turn will type to the `uint256` type when evaluated in the context of a function call.
- `New` is a special typing that's used as the result of applying the `new` operator, and will type to a contract reference when evaluated as a function call.
- `This` and `Super` both refer to the typing of the `this` and `super` keywords and they both alter the resolution rules when used in a member access expression, or type to an `address` type equivalent when used as a value.

The PR also implements some refactors as suggested in the code review of the previous new binder PR #1376:
- re-exporting `NodeId` from instead of defining a new type
- changed import resolver API to work on `NodeId`s instead of `Cursor`s
- refactoring of `binder.rs` module and separating it into several sub-modules
- built-ins related functionality now lives in its own module

### CODE REVIEW

File: crates/codegen/testing/src/common.rs [hunk 1]
```diff
@@ -17,7 +17,14 @@ pub(crate) fn collect_snapshot_tests(
     for file in FileWalker::from_directory(data_dir).find_all()? {
         if let Ok(generated_dir) = file.generated_dir() {
             assert!(
-                generated_dir.unwrap_parent().join("input.sol").exists(),
+                generated_dir.unwrap_parent().join("input.sol").exists()
+                    // handle binder outputs which are placed inside a `binder`
```
Comment (OmarTawfik) [inline crates/codegen/testing/src/common.rs:21]: WDYT of calling the new subfolder `v2` instead of just `binder`? because the whole test suite is already under a `bindings_output` parent folder..

File: crates/solidity/outputs/cargo/tests/src/binder/runner.rs [hunk 2]
```diff
@@ -19,12 +18,7 @@ pub(crate) fn run(group_name: &str, test_name: &str) -> Result<()> {
...
         .join(test_name);
-    let target_dir = CargoWorkspace::locate_source_crate("solidity_testing_snapshots")?
-        .join("target/binder")
-        .join(group_name)
-        .join(test_name)
-        .join("generated");
-    fs::create_dir_all(&target_dir)?;
+    let mut fs = CodegenFileSystem::default();
 
     let compare_with_binding_graph = std::env::var("__SLANG_NEW_BINDER_COMPARE__").is_ok();
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: what is the success rate for this comparison now? are we 100% yet?
If so, should we flip the switch, and make it turned on by default?

File: crates/solidity/outputs/cargo/crate/src/backend/binder/definitions.rs [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+}
+
+#[derive(Debug)]
+pub struct YulVariableDefinition {
+    pub node_id: NodeId,
+    pub identifier: Rc<TerminalNode>,
+}
+
+#[derive(Debug)]
+pub enum Definition {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/definitions.rs]: nit: I suggest putting the `enum Definition` first at the top of the file, since everything else flows from it..

Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/definitions.rs]: Same for `pub(crate) enum Scope` in `scopes.rs`.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    fn new(node_id: NodeId, parent_scope_id: ScopeId) -> Self {
+        Self {
+            node_id,
+            parent_scope_id,
+            definitions: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn insert_definition(&mut self, definition: &Definition) {
+        let symbol = definition.identifier().unparse();
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:206]: There are a few places like here where we unparse/clone a `String` directly, rather than storing an `Rc<>` to the terminal node.

I wonder if this is intended? would this have an impact on perf?

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+        Self {
+            node_id,
+            file_id: file_id.to_string(),
+            definitions: HashMap::new(),
+            imported_files: HashSet::new(),
+            using_directives: Vec::new(),
+        }
+    }
+
+    pub(crate) fn insert_definition(&mut self, definition: &Definition) {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:259]: the whole parent module should be private/never exposed publicly.
Are the `pub(crate)` used everywhere needed?

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+        let node_id = definition.node_id();
+        if let Some(definitions) = self.definitions.get_mut(&symbol) {
+            definitions.push(node_id);
+        } else {
+            self.definitions.insert(symbol, vec![node_id]);
+        }
+    }
+
+    pub(crate) fn add_imported_file(&mut self, file_id: String) {
+        self.imported_files.insert(file_id);
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:270]: These properties are all `pub`, even though they are already (should be?) wrapped by utils like this.
Should they all be turned private?

Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:270]: Same for all other similar types in the parent module..

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+    }
+
+    pub(super) fn lookup_symbol(&self, symbol: &str) -> Vec<NodeId> {
+        self.definitions.get(symbol).cloned().unwrap_or(Vec::new())
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs]: lookup methods like this are on the hot-path, and it seems like they are all cloning or creating black vectors on every call.
I wonder if these methods can just return `impl Iterator<Item = NodeId>`?

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+    }
+
+    pub(crate) fn get_using_directives_in_scope(&self, scope_id: ScopeId) -> Vec<&UsingDirective> {
+        self.get_scope_by_id(scope_id).get_using_directives()
+    }
+
+    pub(crate) fn get_using_directives_in_scope_including_inherited(
+        &self,
+        scope_id: ScopeId,
+    ) -> Vec<&UsingDirective> {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs]: Similar suggestion here. Should all of these methods return an iterator instead of cloning/allocating vectors on every call on the hotpath?

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+        if self
+            .linearisations
+            .insert(node_id, linearised_bases)
+            .is_some()
+        {
+            unreachable!("Trying to linearised twice {node_id:?}");
+        }
+    }
+
+    #[cfg(feature = "__private_backend_api")]
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs]: isn't this flag applied on the entire parent module?

File: crates/solidity/outputs/cargo/crate/src/backend/c3/mod.rs [hunk 10]
```diff
@@ -0,0 +1,239 @@
+use std::collections::{HashMap, VecDeque};
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/c3/mod.rs]: nit: should this module live under `p3_linearise_contracts` module?

File: crates/solidity/outputs/cargo/crate/src/backend/built_ins.rs [hunk 11]
```diff
@@ -0,0 +1,814 @@
+use semver::Version;
+
+use super::binder::{Binder, Definition, Typing};
+use super::types::{Type, TypeId, TypeRegistry};
+use crate::cst::NodeId;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum BuiltIn {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/built_ins.rs:8]: Q: is this module maintained by hand? is it possible to generate these from the language definition?

File: crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs [hunk 12]
```diff
@@ -0,0 +1,1337 @@
...
+        Self {
+            language_version: language_version.clone(),
+            scope_stack: Vec::new(),
+            binder,
+            types,
+        }
+    }
+
+    fn built_ins_resolver(&self) -> BuiltInsResolver<'_> {
+        BuiltInsResolver::new(self.language_version.clone(), &self.binder, &self.types)
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs:73]: similarly: is it possible to create one `BuiltInsResolver` and use it everywhere, instead of cloning/initializing one on every call in the hotpath?

File: crates/solidity/outputs/cargo/crate/src/backend/built_ins.rs (hunk 11 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/built_ins.rs:8]: In that case, we might not need the manual/special type entries in `struct TypeRegistry`, since they will be registered automatically here..

File: crates/solidity/outputs/cargo/tests/Cargo.toml [hunk 13]
```diff
@@ -22,6 +22,7 @@ slang_solidity = { workspace = true, features = [
   "__private_ariadne_errors",
   "__private_testing_utils",
   "__private_backend_api",
+  "__private_wasm_apis",
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/tests/Cargo.toml]: is this still needed? the `CompilationBuilder` from the public API should now be usable..

File: crates/solidity/outputs/cargo/crate/src/backend/mod.rs [hunk 14]
```diff
@@ -1,8 +1,11 @@
 #![allow(missing_docs)]
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/mod.rs]: I suggest copying the documentation from this PR description to the specific types and variants they mention, adding a quick example on how each variant is used.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs [hunk 15]
```diff
@@ -0,0 +1,525 @@
...
+// here, as a typing variant? That has the additional benefit of not requiring
+// us to register every tuple combination.
+#[derive(Clone, Debug)]
+pub enum Typing {
+    Unresolved,
+    Resolved(TypeId),
+    Undetermined(Vec<TypeId>),
+    UserMetaType(NodeId),
+    MetaType(Type),
+    BuiltIn(BuiltIn),
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs:62]: > - `MetaType` similar to `UserMetaType`, but refers to the meta type of an elementary type. A typical use case is explicit casting, which parses as a function call.

Q: Shouldn't the cast expression then be typed as a function call, referencing the type of the built-in?

Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs:62]: > - `BuiltIn` is used to type the `BuiltIn` resolution when the result is not yet an actual type representable in the EVM. Eg. the built-in function `addmod` resolves to `Resolution::BuiltIn(Addmod)` and will type to `Typing::BuiltIn(Addmod)` which in turn will type to the `uint256` type when evaluated in the context of a function call.

Similar question for built-in functions...

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+    New(TypeId),
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs]: nit: I suggest renaming the `New` variant to something like `NewExpression` for clarity..

File: crates/solidity/outputs/cargo/tests/src/binder/runner.rs (hunk 2 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: I'm asking because of the hundreds of newly added snapshots. I'm not sure if manually reviewing them would catch anything between the cracks..

**Review by OmarTawfik — CHANGES_REQUESTED** [review]

File: crates/solidity/outputs/cargo/tests/src/binder/runner.rs (hunk 2 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: I ran some numbers for this PR (I'll report later what results I get in the next one). We get differences in 47 out of 228 snapshots. Of those 47, only in 3 cases there are differences for *all* versions, and in all those 3 cases, the difference is that the stack graph binder cannot disambiguate the definitions in some reference, but the new binder does.

In the remaining 44 cases, there are differences but there is at least one Solidity version where the results are identical. Where differences occur, from what I could quickly gather we have these cases:
- a parsing error prevents construction of some part of the AST, so there are definitions that the new binder cannot see, and hence, cannot resolve
- a built-in binds in the new binder, but not the stack graph; these may be actual bugs in the new binder due to incorrect version checking for some of the built-ins
- ambiguous resolutions that the stack graph binder cannot resolve, but the new binder does; this category will go up in the next PR where we complete the overloads resolution
- other cases not involving built-ins where we're not properly doing version checks in the new binder (eg. modifiers are forbidden in interfaces in >= 0.8.8); I would classify this as a validation issue
- `revert Foo()` statements are parsed as a variable declarations in <0.8.4, but the stack graph binder can still resolve `revert` to the built-in, even though it makes no sense

In no case the stack graph binder and the new binder resolve a reference to a different definition.

I'm not sure about enabling this by default, at least not with the current output. For the development workflow running the comparison slows down the test run times by a factor of 9. I'd usually run just the `binder` tests and they finish in 12 secs, whereas enabling the comparison they take 110 secs. Maybe we can output the differences report in a separate folder for each test? So have, `generated`, `v2/generated`, `diff/generated`. And to avoid slowing down test times, we could generated the `diff` report when running the stack graph snapshots. It's backwards from a logical perspective, but it makes sense, since for those we're already paying the time cost of building the stack graphs.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 4 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:206]: In this case (and the rest of the scopes), we need the string to use it as the hash index for lookup. I tried to be careful with allocations, but maybe some cases slipped through.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 5 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:259]: Probably not. We'll need to tighten the visibility when marking this API as public. I know for a fact that eg. most fields of internal structs are public but they don't need to be. I'm not sure we want to spend the time to clean this up now, or at the next PR. The next PR will introduce a couple of new structs and functions, so it may be worth holding up on this task for now.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 6 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:270]: Yes, similar response to before. I need to review and overhaul all the visibility rules, probably add some helper methods to avoid pub fields, etc.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 7 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs]: Agree, I'll take a closer look. In most cases though I think we will end up using the returned vectors to create a `Resolution::Ambiguous(defs)` with the vector inside. It may be worth delaying the collection up to the last minute though.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs (hunk 8 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs]: This is a different case because the result does not need to be stored, and it makes perfect sense to return an iterator.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs (hunk 9 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs]: Probably a left over when the module was in a different place. I'll remove it.

File: crates/solidity/outputs/cargo/crate/src/backend/c3/mod.rs (hunk 10 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/c3/mod.rs]: Makes sense.

File: crates/solidity/outputs/cargo/crate/src/backend/built_ins.rs (hunk 11 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/built_ins.rs:8]: Yes, this module is hand-written. I'm not sure it'll be possible to generate from the language definition, at least not entirely. We could possibly generate the bulk of it though.

As for the manual entries in the type registry, that'd be true for most. There are a couple of native types that are used in the passes directly (eg. to type logical expressions). In any case, as we discussed offline, I'd prefer to not have to manually register any of the elementary types but instead find a way to encode the type in the type ID. But that's a bigger change.

File: crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs (hunk 12 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs:73]: I don't think it's possible, since we're taking non-mutable references to the binder and types, which would prevent mutation everywhere else. The `BuiltInsResolver` was a convenience struct to avoid having to pass the binder and type registry to all the built-ins functions. Constructing it should be cheap. What we could do is take the language version by reference as well.

File: crates/solidity/outputs/cargo/tests/Cargo.toml (hunk 13 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/Cargo.toml]: Probably a left-over, I'll take a closer look.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs (hunk 15 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs:62]: For cast expressions we could type the casting operator as a function but it's a bit more expensive because we'd need to register the type. More importantly there are some interactions (eg. with the `new` operator) which would make it more cumbersome to work with a function type.

Same thing with the built-ins. We'll need to register a function type for each used built-in function, and there are some edge cases where that's not possible (most prominently, `abi.decode`).

File: crates/solidity/outputs/cargo/tests/src/binder/runner.rs (hunk 2 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: I ran the same comparison in the `beta/binder-2-overloading` branch, which contains the bleeding edge implementation of the new binder.

We have 260 test cases (+32), differences in 75 of them (+28), of which 27 (+24) differ in all versions. In those 27, most of them are cases for ambiguous resolutions that the new binder can resolve but the stack graph cannot (there are many new snapshots that test this functionality). Interestingly, there are 2 cases where the stack graph binder resolves incorrectly (unambiguously) but the new binder does it correctly.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 7 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs]: I changed this instance to return an iterator, but upon further inspection I don't think changing other places would have an impact on performance (not even sure this change has any significant impact). I'd rather wait for the code to stabilize and have hard performance numbers from running callgrind. There are some small inefficiencies which we may want to address later. Keep in mind the only scopes that have vectors are contracts and files.
- in cases where there are no overloads, the vectors will contain a single element that will be cloned and immediately dropped upon creating the `Resolution::Resolved(id)`
- when resolving at the file scope, the initial resolution will be converted to a vector (even if a single result) and fed into a working set (another vector) for expanding imported deconstructed symbols; this needs to happen recursively

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs (hunk 9 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs]: This should be `__private_testing_utils` instead, as this function needs to be exposed to tests only.

File: crates/solidity/outputs/cargo/tests/Cargo.toml (hunk 13 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/Cargo.toml]: Right, the tests are not yet migrated to the public API. I'll work on this.

File: crates/codegen/testing/src/common.rs (hunk 1 above)
Comment (ggiraldez) [reply crates/codegen/testing/src/common.rs:21]: I changed the output folder to `v2` and additionally added the parse errors to the report output.

File: crates/solidity/outputs/cargo/tests/src/binder/runner.rs (hunk 2 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: I implemented these changes:
- Enabled diff reports by default when running the `binder` tests. Having the environment variable `__SLANG_NEW_BINDER_SKIP_DIFF__` set will disable rendering them (and make the test execution a lot 9x faster).
- Extracted the diff report to a different file that gets written in a `diff` folder under each binding snapshot case. The name of the file reflects if we found differences (`-diff` suffix) or not (`-same` suffix).

Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: I'll go over the difference files in detail to see if there is any bug to fix.

Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: I fixed some Solidity version checks we had different in the stack graph vs. built-ins. The remaining differences are due mostly to parsing errors hiding some parts of the CST to the new binder, and a couple related to validation, eg. in interfaces, modifiers are not allowed in >=0.8.8 and `using` directives are not allowed in >=0.7.1. I also fixed some incorrect version checks in the binding rules for the stack graph (eg. `type()` expressions for enums are only allowed in >=0.8.8).

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 5 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:259]: I restricted some functions and fields to `pub(crate)` but there may still be some remaining issues. Scopes are private, but references and definitions are not. For definitions in particular, we may want to reduce the number of "detail" structs, as most of them contain the same fields (node_id + identifier).

File: crates/solidity/outputs/cargo/tests/src/binder/runner.rs (hunk 2 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: Thanks for the thorough review!

Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: Would it be helpful to add a `TODO` with the remaining work items to do/fix once the migration is complete?

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 7 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs]: Sounds good on waiting on stabilization.
Should we add a `TODO` comment with these notes for later?

File: crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs (hunk 15 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/mod.rs:62]: Thanks for explaining.

File: crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs (hunk 12 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs:73]: > take the language version by reference as well.

Makes sense, given the other refs with the same lifetime.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 5 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:259]: My assumption is that `binder` will not be exposed. Binding information would be exposed through the AST nodes (L3/L4?) where they would have links to other AST nodes that are bound to them.

Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs:259]: That's a good point. Yes, in that case both `Binder` and most probably `TypeRegistry` as well will be internal to the backend.

File: crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs (hunk 7 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/crate/src/backend/binder/scopes.rs]: Sounds good, I'll add some `TODO`s.

File: crates/solidity/outputs/cargo/tests/src/binder/runner.rs (hunk 2 above)
Comment (ggiraldez) [reply crates/solidity/outputs/cargo/tests/src/binder/runner.rs]: I'll add some comments. Since the changes introduced in the overloads+correctness branch, the bindings will diverge even more from the stack graph, because we are able to resolve more ambiguities. So I think we should do another thorough review of the differences in the next PR.
